File: crates/sky-cua-service/src/phone/companion/protocol.rs
//! Companion RPC wire DTOs (protocol_version = 1).
//!
//! These types are the serde mirror of the JSON contract documented in
//! `docs/runtime/phone-companion-protocol.md`. Transport is HTTP/1.1
//! `POST /rpc` to a host-managed `adb forward tcp:PORT tcp:PORT` localhost
//! endpoint inside the companion app: one JSON request -> one JSON response.
//!
//! Request:  `{"protocol_version":1,"token":"<tok>","id":<int>,"method":"<m>","params":{...}}`
//! Response (ok):    `{"protocol_version":1,"ok":true,"id":<int>,"result":{...}}`
//! Response (error): `{"protocol_version":1,"ok":false,"id":<int>,"error":{"code":"<c>","message":"<m>"}}`
//!
//! Every DTO carries `pub(crate)` visibility so the integrator (`manager.rs`)
//! can name the request/result types when it wires the client into routing.
//!
//! Until the integrator wires the client into `manager.rs`, these DTOs are only
//! referenced from the client/tests. The module-level expectation keeps non-test
//! builds clean and matches the spine's `expect(dead_code)` idiom; it becomes
//! unfulfilled (and self-removing) once routing references the types.
#![cfg_attr(not(test), expect(dead_code))]

use serde::{Deserialize, Serialize};

/// The only protocol version this host speaks. Sent on every request; the
/// companion rejects mismatches with [`error_codes::VERSION_MISMATCH`].
pub(crate) const PROTOCOL_VERSION: u32 = 1;

/// Wire-level RPC method names. Kept as constants so the client and tests never
/// drift on a string literal.
pub(crate) mod methods {
    pub(crate) const HEALTH: &str = "health";
    pub(crate) const CAPABILITIES: &str = "capabilities";
    pub(crate) const ACCESSIBILITY_TREE: &str = "accessibility_tree";
    pub(crate) const SCREENSHOT: &str = "screenshot";
    pub(crate) const GESTURE: &str = "gesture";
    pub(crate) const CURSOR_OVERLAY: &str = "cursor_overlay";
    pub(crate) const OVERLAY_ACTIVE: &str = "overlay_active";
    pub(crate) const OVERLAY_GESTURE: &str = "overlay_gesture";
    pub(crate) const NOTIFICATIONS: &str = "notifications";
    pub(crate) const NOTIFICATION_OP: &str = "notification_op";
    pub(crate) const CURRENT_APP: &str = "current_app";
    pub(crate) const APP_LIST: &str = "app_list";
    pub(crate) const APP_OP: &str = "app_op";
}

/// Per-method error codes the companion may return in `error.code`, plus the
/// protocol-level codes (`unauthorized`, `version_mismatch`) the server applies
/// before method dispatch. Clients route on these stable codes, never prose.
pub(crate) mod error_codes {
    // Protocol-level (validated before method dispatch). `unauthorized` and
    // `version_mismatch` map to their own host variants. `unknown_method` and
    // `internal` signal the request never reached a working method handler, so
    // the host falls back rather than treating them as per-method application
    // errors. `bad_request` is overloaded (dispatch-level validation AND genuine
    // per-method app errors such as a bad intent URI), indistinguishable on the
    // wire, so the host does NOT fall back on it — it is a non-fallback Method
    // error and the affected action falls back on its own.
    pub(crate) const UNAUTHORIZED: &str = "unauthorized";
    pub(crate) const VERSION_MISMATCH: &str = "version_mismatch";
    pub(crate) const BAD_REQUEST: &str = "bad_request";
    pub(crate) const UNKNOWN_METHOD: &str = "unknown_method";
    pub(crate) const INTERNAL: &str = "internal";

    // screenshot method.
    pub(crate) const SECURE_WINDOW: &str = "secure_window";
    pub(crate) const UNSUPPORTED_API: &str = "unsupported_api";
    pub(crate) const DISABLED_SERVICE: &str = "disabled_service";
    pub(crate) const OEM_POLICY: &str = "oem_policy";
    pub(crate) const THROTTLED: &str = "throttled";
    pub(crate) const TRANSIENT: &str = "transient";

    // notification_op method.
    pub(crate) const GONE: &str = "gone";
    pub(crate) const REDACTED: &str = "redacted";
    pub(crate) const PENDING_INTENT_MISSING: &str = "pending_intent_missing";
    pub(crate) const CANCELED: &str = "canceled";
    pub(crate) const EXPIRED: &str = "expired";
    pub(crate) const IMMUTABLE: &str = "immutable";
    pub(crate) const REPLY_UNAVAILABLE: &str = "reply_unavailable";
    pub(crate) const OEM_FILTERED: &str = "oem_filtered";
}

// ===========================================================================
// Envelopes
// ===========================================================================

/// Outgoing request envelope. `params` is a typed payload that serializes to the
/// `params` object; methods with no params use [`NoParams`].
#[derive(Debug, Clone, Serialize, PartialEq)]
pub(crate) struct RpcRequest<P> {
    pub(crate) protocol_version: u32,
    pub(crate) token: String,
    pub(crate) id: u64,
    pub(crate) method: String,
    pub(crate) params: P,
}

impl<P> RpcRequest<P> {
    /// Build a v1 request for `method` with `token`, `id`, and typed `params`.
    pub(crate) fn new(token: impl Into<String>, id: u64, method: &str, params: P) -> Self {
        Self {
            protocol_version: PROTOCOL_VERSION,
            token: token.into(),
            id,
            method: method.to_string(),
            params,
        }
    }
}

/// Empty params payload (serializes to `{}`). Used by `health`, `capabilities`,
/// and `current_app`.
#[derive(Debug, Clone, Default, Serialize, Deserialize, PartialEq, Eq)]
pub(crate) struct NoParams {}

/// The structured `error` object on a non-ok response.
#[derive(Debug, Clone, Serialize, Deserialize, PartialEq, Eq)]
pub(crate) struct RpcError {
    pub(crate) code: String,
    pub(crate) message: String,
}

/// Raw response envelope before result typing. `result` stays as a
/// `serde_json::Value` so one decode pass validates the envelope, then the
/// client decodes `result` into the per-method DTO. This lets malformed-result
/// payloads surface as a typed decode failure rather than an envelope failure.
#[derive(Debug, Clone, Deserialize, PartialEq)]
pub(crate) struct RpcEnvelope {
    pub(crate) protocol_version: u32,
    pub(crate) ok: bool,
    pub(crate) id: u64,
    #[serde(default)]
    pub(crate) result: Option<serde_json::Value>,
    #[serde(default)]
    pub(crate) error: Option<RpcError>,
}

// ===========================================================================
// health / capabilities
// ===========================================================================

/// `health` result. `capabilities` returns a superset (see [`CapabilitiesResult`]).
#[derive(Debug, Clone, Serialize, Deserialize, PartialEq, Eq)]
pub(crate) struct HealthResult {
    pub(crate) version: String,
    pub(crate) version_code: u64,
    pub(crate) package: String,
    pub(crate) accessibility_enabled: bool,
    pub(crate) can_perform_gestures: bool,
    pub(crate) can_retrieve_window_content: bool,
    pub(crate) can_take_screenshot: bool,
    pub(crate) notification_listener_enabled: bool,
    pub(crate) native_overlay: bool,
    pub(crate) native_overlay_pass_through: bool,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub(crate) privileged_setup: Option<String>,
}

/// `capabilities` result: the health fields plus screenshot/gesture support
/// detail. Flattening `health` keeps the wire object a single flat map while the
/// Rust type composes the shared fields.
#[derive(Debug, Clone, Serialize, Deserialize, PartialEq, Eq)]
pub(crate) struct CapabilitiesResult {
    #[serde(flatten)]
    pub(crate) health: HealthResult,
    pub(crate) screenshot_api_level: u32,
    pub(crate) screenshot_supported: bool,
    pub(crate) gesture_supported: bool,
}

// ===========================================================================
// accessibility_tree
// ===========================================================================

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq, Eq)]
pub(crate) struct AccessibilityTreeParams {
    pub(crate) max_nodes: u32,
}

/// One node in the companion's flat accessibility list. `bounds` is the raw
/// `[left, top, right, bottom]` device-pixel rect the companion reports.
#[derive(Debug, Clone, Serialize, Deserialize, PartialEq, Eq)]
pub(crate) struct AccessibilityNodeDto {
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub(crate) class: Option<String>,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub(crate) text: Option<String>,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub(crate) content_desc: Option<String>,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub(crate) bounds: Option<[i32; 4]>,
    pub(crate) focusable: bool,
    pub(crate) enabled: bool,
    pub(crate) clickable: bool,
}

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq, Eq)]
pub(crate) struct AccessibilityTreeResult {
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub(crate) package: Option<String>,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub(crate) activity: Option<String>,
    #[serde(default)]
    pub(crate) nodes: Vec<AccessibilityNodeDto>,
    pub(crate) truncated: bool,
    pub(crate) redacted: bool,
}

// ===========================================================================
// screenshot
// ===========================================================================

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq, Eq)]
pub(crate) struct ScreenshotParams {
    pub(crate) include_overlay: bool,
}

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq, Eq)]
pub(crate) struct ScreenshotResult {
    pub(crate) mime_type: String,
    pub(crate) data_base64: String,
    pub(crate) width: u32,
    pub(crate) height: u32,
    pub(crate) contains_native_overlay: bool,
}

// ===========================================================================
// gesture
// ===========================================================================

/// Gesture kind. `tap` uses one point; `swipe` uses two (start, end).
#[derive(Debug, Clone, Copy, Serialize, Deserialize, PartialEq, Eq)]
#[serde(rename_all = "snake_case")]
pub(crate) enum GestureKind {
    Tap,
    Swipe,
}

/// A device-pixel point in a gesture path.
#[derive(Debug, Clone, Copy, Serialize, Deserialize, PartialEq)]
pub(crate) struct GesturePoint {
    pub(crate) x: f64,
    pub(crate) y: f64,
}

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq)]
pub(crate) struct GestureParams {
    pub(crate) kind: GestureKind,
    pub(crate) points: Vec<GesturePoint>,
    pub(crate) duration_ms: u32,
}

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq, Eq)]
pub(crate) struct GestureResult {
    pub(crate) dispatched: bool,
}

// ===========================================================================
// cursor_overlay
// ===========================================================================

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq)]
pub(crate) struct CursorOverlayParams {
    pub(crate) visible: bool,
    pub(crate) x: f64,
    pub(crate) y: f64,
}

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq, Eq)]
pub(crate) struct CursorOverlayResult {
    pub(crate) shown: bool,
    pub(crate) pass_through: bool,
}

// ===========================================================================
// overlay_active
// ===========================================================================

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq, Eq)]
pub(crate) struct OverlayActiveParams {
    pub(crate) active: bool,
}

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq, Eq)]
pub(crate) struct OverlayActiveResult {
    pub(crate) active: bool,
    pub(crate) glow_supported: bool,
}

// ===========================================================================
// overlay_gesture
// ===========================================================================

/// Reuses [`GesturePoint`] for the device-pixel path. `kind` is the free-form
/// wire string (`tap`/`swipe`/`drag`) rather than [`GestureKind`], because the
/// visual overlay supports `drag`, which the real-input `gesture` method does
/// not.
#[derive(Debug, Clone, Serialize, Deserialize, PartialEq)]
pub(crate) struct OverlayGestureParams {
    pub(crate) kind: String,
    pub(crate) points: Vec<GesturePoint>,
    pub(crate) duration_ms: u32,
}

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq, Eq)]
pub(crate) struct OverlayGestureResult {
    pub(crate) animated: bool,
}

// ===========================================================================
// notifications / notification_op
// ===========================================================================

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq, Eq)]
pub(crate) struct NotificationsParams {
    pub(crate) max: u32,
}

/// Redaction state of a companion notification event.
#[derive(Debug, Clone, Copy, Serialize, Deserialize, PartialEq, Eq)]
#[serde(rename_all = "snake_case")]
pub(crate) enum NotificationRedactionDto {
    None,
    Partial,
    Full,
}

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq, Eq)]
pub(crate) struct NotificationActionDto {
    pub(crate) action_id: String,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub(crate) title: Option<String>,
    pub(crate) is_reply: bool,
}

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq, Eq)]
pub(crate) struct NotificationEventDto {
    pub(crate) event_id: String,
    pub(crate) package: String,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub(crate) channel: Option<String>,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub(crate) title: Option<String>,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub(crate) body: Option<String>,
    pub(crate) redaction: NotificationRedactionDto,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub(crate) ranking: Option<u32>,
    /// Whether the notification carries a content-intent the agent can open.
    /// Absent on older companions; the host defaults it conservatively.
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub(crate) can_open: Option<bool>,
    /// Whether the notification is user-dismissable (`StatusBarNotification.isClearable`).
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub(crate) can_dismiss: Option<bool>,
    /// Whether the notification is an ongoing/non-clearable event.
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub(crate) ongoing: Option<bool>,
    pub(crate) when_ms: u64,
    #[serde(default)]
    pub(crate) actions: Vec<NotificationActionDto>,
}

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq, Eq)]
pub(crate) struct NotificationsResult {
    pub(crate) listener_enabled: bool,
    #[serde(default)]
    pub(crate) events: Vec<NotificationEventDto>,
    pub(crate) truncated: bool,
}

/// `notification_op` operation kind.
#[derive(Debug, Clone, Copy, Serialize, Deserialize, PartialEq, Eq)]
#[serde(rename_all = "snake_case")]
pub(crate) enum NotificationOp {
    Open,
    Dismiss,
    Action,
    Reply,
}

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq, Eq)]
pub(crate) struct NotificationOpParams {
    pub(crate) event_id: String,
    pub(crate) op: NotificationOp,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub(crate) action_id: Option<String>,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub(crate) reply_text: Option<String>,
}

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq, Eq)]
pub(crate) struct NotificationOpResult {
    pub(crate) ok: bool,
}

// ===========================================================================
// current_app / app_list / app_op
// ===========================================================================

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq, Eq)]
pub(crate) struct CurrentAppResult {
    pub(crate) package: String,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub(crate) activity: Option<String>,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub(crate) label: Option<String>,
}

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq, Eq)]
pub(crate) struct AppListParams {
    pub(crate) launchable_only: bool,
}

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq, Eq)]
pub(crate) struct AppListEntryDto {
    pub(crate) package: String,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub(crate) label: Option<String>,
    pub(crate) launchable: bool,
}

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq, Eq)]
pub(crate) struct AppListResult {
    #[serde(default)]
    pub(crate) apps: Vec<AppListEntryDto>,
    pub(crate) truncated: bool,
}

/// `app_op` operation kind.
#[derive(Debug, Clone, Copy, Serialize, Deserialize, PartialEq, Eq)]
#[serde(rename_all = "snake_case")]
pub(crate) enum AppOp {
    Launch,
    OpenIntent,
    ForceStop,
}

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq, Eq)]
pub(crate) struct AppOpParams {
    pub(crate) op: AppOp,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub(crate) package: Option<String>,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub(crate) intent_uri: Option<String>,
}

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq, Eq)]
pub(crate) struct AppOpResult {
    pub(crate) ok: bool,
}

File: android/phone-companion/app/src/main/java/com/skycua/phonecompanion/protocol/Protocol.kt
package com.skycua.phonecompanion.protocol

import com.skycua.phonecompanion.json.JsonParseException
import com.skycua.phonecompanion.json.JsonParser
import com.skycua.phonecompanion.json.JsonValue
import com.skycua.phonecompanion.json.JsonWriter
import com.skycua.phonecompanion.json.jsonObject

/**
 * Wire protocol contract for the Sky Phone Companion RPC, mirroring
 * docs/runtime/phone-companion-protocol.md. This file is the single source of
 * truth on the Android side for envelope shapes, the protocol version, method
 * names, and error codes.
 */
object Protocol {
    const val VERSION: Long = 1

    object Methods {
        const val HEALTH = "health"
        const val CAPABILITIES = "capabilities"
        const val ACCESSIBILITY_TREE = "accessibility_tree"
        const val SCREENSHOT = "screenshot"
        const val APP_SHOT = "appshot"
        const val GESTURE = "gesture"
        const val CURSOR_OVERLAY = "cursor_overlay"
        const val OVERLAY_ACTIVE = "overlay_active"
        const val OVERLAY_GESTURE = "overlay_gesture"
        const val NOTIFICATIONS = "notifications"
        const val NOTIFICATION_OP = "notification_op"
        const val CURRENT_APP = "current_app"
        const val APP_LIST = "app_list"
        const val APP_OP = "app_op"
        const val CLIPBOARD = "phone_clipboard"
        const val EDITOR = "phone_editor"
        const val STORAGE = "phone_storage"
        const val CAMERA = "phone_camera"
        const val KEY = "phone_key"
        /** Direct-only phone-control.v2 method; intentionally absent from ALL. */
        const val SMS_QUERY = "sms.query"

        val ALL =
            setOf(
                HEALTH,
                CAPABILITIES,
                ACCESSIBILITY_TREE,
                SCREENSHOT,
                APP_SHOT,
                GESTURE,
                CURSOR_OVERLAY,
                OVERLAY_ACTIVE,
                OVERLAY_GESTURE,
                NOTIFICATIONS,
                NOTIFICATION_OP,
                CURRENT_APP,
                APP_LIST,
                APP_OP,
                CLIPBOARD,
                EDITOR,
                STORAGE,
                CAMERA,
                KEY,
            )
    }

    /**
     * Error codes the companion may emit. Envelope/transport-level codes
     * ([UNAUTHORIZED], [VERSION_MISMATCH], [UNKNOWN_METHOD], [BAD_REQUEST]) are
     * distinct from per-method application errors (screenshot/notification
     * codes), which the host treats as successful RPCs reporting an operation
     * could not be performed.
     */
    object ErrorCodes {
        // Envelope / auth / dispatch.
        const val UNAUTHORIZED = "unauthorized"
        const val VERSION_MISMATCH = "version_mismatch"
        const val UNKNOWN_METHOD = "unknown_method"
        const val BAD_REQUEST = "bad_request"
        const val INTERNAL = "internal"

        // Screenshot application errors.
        const val SECURE_WINDOW = "secure_window"
        const val UNSUPPORTED_API = "unsupported_api"
        const val DISABLED_SERVICE = "disabled_service"

        // Reserved for the screenshot route but currently unreachable: the
        // AccessibilityService screenshot callback exposes no OEM-policy signal,
        // so no path emits this code. Kept in the contract for forward
        // compatibility (mirrored note in the Rust protocol doc).
        const val OEM_POLICY = "oem_policy"
        const val THROTTLED = "throttled"
        const val TRANSIENT = "transient"

        // notification_op application errors.
        const val GONE = "gone"
        const val REDACTED = "redacted"
        const val PENDING_INTENT_MISSING = "pending_intent_missing"
        const val CANCELED = "canceled"
        const val EXPIRED = "expired"
        const val IMMUTABLE = "immutable"
        const val REPLY_UNAVAILABLE = "reply_unavailable"
        const val OEM_FILTERED = "oem_filtered"
    }
}

/** A decoded request envelope. */
data class RpcRequest(
    val protocolVersion: Long,
    val token: String?,
    val id: Long,
    val method: String,
    val params: JsonValue.Obj,
)

/** A decoded/encoded response envelope. */
sealed class RpcResponse {
    abstract val id: Long

    data class Success(
        override val id: Long,
        val result: JsonValue.Obj,
    ) : RpcResponse()

    data class Failure(
        override val id: Long,
        val code: String,
        val message: String,
    ) : RpcResponse()
}

/** Thrown when an incoming envelope cannot be parsed into an [RpcRequest]. */
class EnvelopeException(
    val code: String,
    override val message: String,
    val id: Long = 0,
) : Exception(message)

/** (De)serialization of request and response envelopes. */
object Envelope {
    /**
     * Parses a request body into an [RpcRequest]. The protocol version is *not*
     * enforced here; the dispatcher checks it after parsing so it can return a
     * structured [Protocol.ErrorCodes.VERSION_MISMATCH] with the request id.
     *
     * @throws EnvelopeException with [Protocol.ErrorCodes.BAD_REQUEST] when the
     *   body is not a JSON object, the id is missing/non-integral, or the method
     *   is missing.
     */
    fun parseRequest(body: String): RpcRequest {
        val obj =
            try {
                JsonParser.parseObject(body)
            } catch (e: JsonParseException) {
                throw EnvelopeException(Protocol.ErrorCodes.BAD_REQUEST, "malformed JSON: ${e.message}")
            }

        val idNum =
            (obj["id"] as? JsonValue.Num)
                ?: throw EnvelopeException(Protocol.ErrorCodes.BAD_REQUEST, "missing or non-numeric id")
        if (!idNum.isIntegral) {
            throw EnvelopeException(Protocol.ErrorCodes.BAD_REQUEST, "id must be an integer")
        }
        val id = idNum.toLong()

        val protocolVersion =
            (obj["protocol_version"] as? JsonValue.Num)?.toLong()
                ?: throw EnvelopeException(
                    Protocol.ErrorCodes.BAD_REQUEST,
                    "missing protocol_version",
                    id,
                )

        val method =
            obj.string("method")
                ?: throw EnvelopeException(Protocol.ErrorCodes.BAD_REQUEST, "missing method", id)

        val token = obj.string("token")
        val params = obj.obj("params") ?: JsonValue.Obj(emptyMap())

        return RpcRequest(
            protocolVersion = protocolVersion,
            token = token,
            id = id,
            method = method,
            params = params,
        )
    }

    fun encodeResponse(response: RpcResponse): String {
        val obj =
            when (response) {
                is RpcResponse.Success ->
                    jsonObject {
                        put("protocol_version", Protocol.VERSION)
                        put("ok", true)
                        put("id", response.id)
                        put("result", response.result)
                    }
                is RpcResponse.Failure ->
                    jsonObject {
                        put("protocol_version", Protocol.VERSION)
                        put("ok", false)
                        put("id", response.id)
                        put(
                            "error",
                            jsonObject {
                                put("code", response.code)
                                put("message", response.message)
                            },
                        )
                    }
            }
        return JsonWriter.write(obj)
    }
}
